Dynamically changing tooltipBgColor

apekshamehta opened the pull request:
**Changes Made:**

In the `bar_chart_data.dart` file:
- Added a new parameter `getTooltipColor` to the constructor of `BarTouchTooltipData`.
- Defined a new type `GetBarTooltipColor`, which is a function that takes a `BarChartGroupData` parameter and returns a `Color`.

In the `bar_chart_painter.dart` file:
- Assigned the tooltip color to the `_bgTouchTooltipPaint.color` property using the `getTooltipColor` function from the `tooltipData` object.

Additionally, the `getTooltipColor` function has been utilized in the `bar_chart_sample1.dart` file within the `feature/dynamic-tooltip-bgcolor` branch.

Please review the changes for this pull request.

imaNNeo commented on lib/src/chart/bar_chart/bar_chart_data.dart:
```diff
@@ -700,6 +700,7 @@ class BarTouchTooltipData with EquatableMixin {
     double? tooltipHorizontalOffset,
     double? maxContentWidth,
     GetBarTooltipItem? getTooltipItem,
+    GetBarTooltipColor? getTooltipColor,
```
What if we remove `tooltipBgColor` completely?

apekshamehta replied:
Yes we could remove it completely but it was kept considering backward compatibility.

apekshamehta replied:
this is done. old api is removed. changelog is updated.

imaNNeo commented:
1. Please read our [contributing guidelines](https://github.com/imaNNeo/fl_chart/blob/master/CONTRIBUTING.md) carefully
2. Please update our [CHANGELOG](https://github.com/imaNNeo/fl_chart/blob/master/CHANGELOG.md)
3. Please update our docs (bar_chart.md)
4. Please do the same for other axis-based charts such as line chart and scatter chart

apekshamehta commented:
> 1. Please read our [contributing guidelines](https://github.com/imaNNeo/fl_chart/blob/master/CONTRIBUTING.md) carefully
> 2. Please update our [CHANGELOG](https://github.com/imaNNeo/fl_chart/blob/master/CHANGELOG.md)
> 3. Please update our docs (bar_chart.md)
> 4. Please do the same for other axis-based charts such as line chart and scatter chart

Ok i am starting with first 3 points. Will continue with 4th point once i get confirmation for the change i have made for the bar chart.

imaNNeo commented:
I've answered

apekshamehta commented:
I have done all the 

> 1. Please read our [contributing guidelines](https://github.com/imaNNeo/fl_chart/blob/master/CONTRIBUTING.md) carefully
> 2. Please update our [CHANGELOG](https://github.com/imaNNeo/fl_chart/blob/master/CHANGELOG.md)
> 3. Please update our docs (bar_chart.md)
> 4. Please do the same for other axis-based charts such as line chart and scatter chart

this is also done. please review and close the pr. also in the meantime did get a good grasp on dart. so open to some exciting new issues that are of priority. thanks

imaNNeo commented on lib/src/chart/bar_chart/bar_chart_data.dart:
```diff
@@ -761,10 +759,13 @@ class BarTouchTooltipData with EquatableMixin {
   @override
   List<Object?> get props => [
-        tooltipBgColor,
+        // tooltipBgColor,
```
Remote comments please

apekshamehta replied:
unnecessary comments are removed please check latest commit.

imaNNeo commented on lib/src/chart/scatter_chart/scatter_chart_painter.dart:
```diff
@@ -320,7 +320,8 @@ class ScatterChartPainter extends AxisChartPainter<ScatterChartData> {
       bottomRight: radius,
     );
-    _bgTouchTooltipPaint.color = tooltipData.tooltipBgColor;
+    // _bgTouchTooltipPaint.color = tooltipData.tooltipBgColor;
```
Remove comment

apekshamehta replied:
Removed comments.

imaNNeo commented on repo_files/documentations/bar_chart.md:
```diff
@@ -106,6 +105,7 @@ enum values {`start`, `end`, `center`, `spaceEvenly`, `spaceAround`, `spaceBetwe
  |fitInsideHorizontally| forces tooltip to horizontally shift inside the chart's bounding box| false|
  |fitInsideVertically| forces tooltip to vertically shift inside the chart's bounding box| false|
  |direction| Controls showing tooltip on top or bottom, default is auto.| auto|
+ |getTooltipColor|a callback that retrieves the Color for each rod separately from the given [BarChartGroupData](#BarChartGroupData) to set the background color of the tooltip bubble|Colors.white| 
```
Default color (last column) is not correct

apekshamehta replied:
Done.

imaNNeo commented on repo_files/documentations/line_chart.md:
```diff
@@ -132,6 +130,7 @@ When you change the chart's state, it animates to the new state internally (usin
  |fitInsideHorizontally| forces tooltip to horizontally shift inside the chart's bounding box| false|
  |fitInsideVertically| forces tooltip to vertically shift inside the chart's bounding box| false|
  |showOnTopOfTheChartBoxArea| forces the tooltip container to top of the line| false|
+ |getTooltipColor|a callback that retrieves the Color for each touched spots separately from the given [LineBarSpot](#LineBarSpot) to set the background color of the tooltip bubble|Colors.white| 
```
Default color (last column) is not correct

apekshamehta replied:
Done.

imaNNeo commented on repo_files/documentations/scatter_chart.md:
```diff
@@ -59,6 +58,7 @@ When you change the chart's state, it animates to the new state internally (usin
 |getTooltipItems|a callback that retrieve a [ScatterTooltipItem](#ScatterTooltipItem) by the given [ScatterSpot](#ScatterSpot) |defaultScatterTooltipItem|
 |fitInsideHorizontally| forces tooltip to horizontally shift inside the chart's bounding box| false|
 |fitInsideVertically| forces tooltip to vertically shift inside the chart's bounding box| false|
+|getTooltipColor|a callback that retrieves the Color for each touched spots separately from the given [ScatterSpot](#ScatterSpot) to set the background color of the tooltip bubble|Colors.white| 
```
Default color (last column) is not correct

apekshamehta replied:
I have corrected it in all the files please check latest commit.

imaNNeo commented on pubspec.yaml:
```diff
@@ -1,7 +1,10 @@
 description: A highly customizable Flutter chart library that supports Line Chart, Bar Chart, Pie Chart, Scatter Chart, and Radar Chart.
-version: 0.62.0
-homepage: https://github.com/imaNNeo/fl_chart
+version: 0.64.0
```
You don't need to increase the version.

apekshamehta replied:
Done.

imaNNeo commented on CHANGELOG.md:
```diff
@@ -1,4 +1,10 @@
-## newVersion
+## nextVersion
+
+## 0.64.0
```
No need to mention a new version, please just put your changes under the ## newVersion section, I will bump the version number when I released a new version.

Thanks!

apekshamehta replied:
ok its done.

imaNNeo commented on CHANGELOG.md:
```diff
@@ -2,6 +2,11 @@
 * **BUGFIX** (by @imaNNeo) Fix Negative BarChartRodStackItem are not drawn correctly bug, #1347
 
+## nextVersion
+* **FEATURE**  (by @apekshamehta) Added new method called getTooltipColor for axis charts (bar,line,scatter) to change backgroud color of tooltip dynamically.[issue](https://github.com/imaNNeo/fl_chart/issues/1279).
```
There is a typo here: `backgroud`

apekshamehta replied:
Solved

imaNNeo commented on CHANGELOG.md:
```diff
@@ -2,6 +2,11 @@
 * **BUGFIX** (by @Anas35) Fix Tooltip not displaying when value from BackgroundBarChartRodData is less than zero. #1345.
 * **BUGFIX** (by @imaNNeo) Fix Negative BarChartRodStackItem are not drawn correctly bug, #1347
 
```
Please remove empty space

imaNNeo commented on example/lib/presentation/samples/bar/bar_chart_sample1.dart:
```diff
@@ -162,7 +162,35 @@ class BarChartSample1State extends State<BarChartSample1> {
-          tooltipBgColor: Colors.blueGrey,
+          getTooltipColor: (group) {
+            Color bgColor;
+            switch (group.x) {
```
Please use [the dart3 switch expression](https://dart.dev/language/branches#switch-expressions) here.
It can be like this:
```dart
Color bgColor = switch(group.x) {
   0 => AppColors.contentColorPurple.withAlpha(100),
   1 => AppColors.contentColorYellow.withAlpha(100),
...
}

apekshamehta replied:
Done

imaNNeo commented on test/chart/line_chart/line_chart_painter_test.dart:
```diff
@@ -2257,7 +2257,7 @@ void main() {
 
       final tooltipData = LineTouchTooltipData(
-        tooltipBgColor: const Color(0x11111111),
+        getTooltipColor: (tochedSpots) => const Color(0x11111111),
```
Typo on `tochedSpots`

apekshamehta replied:
I have solved it in my recent commit

imaNNeo commented on test/chart/line_chart/line_chart_painter_test.dart:
```diff
@@ -2367,7 +2367,7 @@ void main() {
 
       final tooltipData = LineTouchTooltipData(
-        tooltipBgColor: const Color(0x11111111),
+        getTooltipColor: (tochedSpots) => const Color(0x11111111),
```
Typo

apekshamehta replied:
Solved.

imaNNeo commented on test/chart/line_chart/line_chart_painter_test.dart:
```diff
@@ -2477,7 +2477,7 @@ void main() {
 
       final tooltipData = LineTouchTooltipData(
-        tooltipBgColor: const Color(0x11111111),
+        getTooltipColor: (tochedSpots) => const Color(0x11111111),
```
Typo

apekshamehta replied:
Solved

imaNNeo commented on lib/src/chart/bar_chart/bar_chart_data.dart:
```diff
@@ -715,16 +714,15 @@ class BarTouchTooltipData with EquatableMixin {
         maxContentWidth = maxContentWidth ?? 120,
         getTooltipItem = getTooltipItem ?? defaultBarTooltipItem,
+        getTooltipColor = getTooltipColor ??
+            ((BarChartGroupData? group) => Colors.blueGrey.darken(15)),
```
To keep consistency, please use a defined function just like `defaultLineTooltipColor` and `defaultScatterTooltipColor` here

apekshamehta replied:
Done

imaNNeo commented:
I think you also need a rebase,
After that, we are ready to merge it.
Thanks in advance!

imaNNeo commented:
I think you need to [rebase](https://git-scm.com/book/en/v2/Git-Branching-Rebasing) your branch.

imaNNeo commented:
You should `rebase` your branch. So when you merge master into your branch you cannot rebase anymore.
Please fix your commits (you can reset your branch and commit your changes again and then you can force push to override the remote commits)

apekshamehta commented:
> You should `rebase` your branch. So when you merge master into your branch you cannot rebase anymore. Please fix your commits (you can reset your branch and commit your changes again and then you can force push to override the remote commits)

@imaNNeo I have rebased my branch feature/dynamic-tooltip-bgcolor. I have created new pull request please check.

imaNNeo commented:
Hi, I'm sorry I think I ruined something here :)
I wanted to rebase your branch and resolve the new conflicts, but I have stolen your commits and everything is in my name.
Can you please create a fresh PR and commit everything to your name?

apekshamehta commented:
> Hi, I'm sorry I think I ruined something here :) I wanted to rebase your branch and resolve the new conflicts, but I have stolen your commits and everything is in my name. Can you please create a fresh PR and commit everything to your name?

Sorry for the delay will complete this by today.

xvrh commented:
@apekshamehta @imaNNeo I'm interested in this feature.
I rebased and squashed the work of this branch here: https://github.com/xvrh/fl_chart/commits/feature/dynamic-tooltip-bgcolor/
How can we move this forward to be merged?

Thanks!

apekshamehta commented:
> @apekshamehta @imaNNeo I'm interested in this feature. I rebased and squashed the work of this branch here: https://github.com/xvrh/fl_chart/commits/feature/dynamic-tooltip-bgcolor/ How can we move this forward to be merged?
> 
> Thanks!

I have tested the changes and they are working as expected let's get this merged

imaNNeo commented:
This PR contains 80 wrong commits (my bad).
Can you please create a new PR with your single commit?

imaNNeo commented:
Don't forget to mention this PR to have it in the history if you created a new one

xvrh commented:
PR created here: https://github.com/imaNNeo/fl_chart/pull/1595
